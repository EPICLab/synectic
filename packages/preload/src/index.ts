/**
 * @module preload
 */

export {fileOpen, fileSave} from './dialogs';
export {flattenArray, flattenObject} from './flatten';
export {
  add,
  catFile,
  checkGitVersion,
  checkRefFormat,
  checkUnmergedBranch,
  checkUnmergedPath,
  checkoutBranch,
  checkoutDetached,
  checkoutNewBranch,
  checkoutPathspec,
  cloneRepo,
  commit,
  copyBranch,
  createBranch,
  deleteBranch,
  errorLogger,
  extractFromURL,
  extractRepoName,
  fetchMergingBranches,
  fileStatus,
  filterIgnore,
  getBranchRoot,
  getConfig,
  getCredentials,
  getIgnore,
  getMatchUp,
  getRemote,
  getRemoteConfig,
  getRoot,
  getWorktreePaths,
  isValidRepository,
  isValidRepositoryURL,
  listBranch,
  log,
  mergeBase,
  mergeBranch,
  mergeInProgress,
  moveBranch,
  processBranchOutput,
  processCatFileOutput,
  processLogOutput,
  processMergeBaseOutput,
  processMergeOutput,
  processRemoteOutput,
  processShowBranchOutput,
  processStatusCode,
  processWorktreeOutput,
  reset,
  resolveURL,
  restore,
  revList,
  revParse,
  revParseArgs,
  setConfig,
  setUpstreamBranch,
  showBranch,
  unsetUpstreamBranch,
  worktreeAdd,
  worktreeList,
  worktreeLock,
  worktreeMove,
  worktreePrune,
  worktreeRemove,
  worktreeRepair,
  worktreeStatus,
  worktreeUnlock,
  type Conflict,
  type GitConfig,
  type MergeAction,
  type WorktreePaths,
} from './git';
export {
  decompressBinaryObject,
  execute,
  extractDirname,
  extractExtension,
  extractFilename,
  extractStats,
  filterReadArray,
  getDescendantPaths,
  isDescendant,
  isDirectory,
  isEqualPaths,
  readDirAsync,
  readDirAsyncDepth,
  readFileAsync,
  replaceExt,
  validateFileName,
  writeFileAsync,
} from './io';
export {
  clipboard,
  cwd,
  dirname,
  join,
  normalize,
  openExternal,
  platform,
  relative,
  sep,
  sha256sum,
  uuid,
} from './platform';
export {
  asyncFilter,
  delay,
  deserialize,
  equalArrayBuffers,
  equalArrays,
  equalMaps,
  filterObject,
  filteredArrayEquality,
  filteredObjectEquality,
  getConflictingChunks,
  getRandomInt,
  hasUpdates,
  isDefined,
  isFilled,
  isModified,
  isNumber,
  isPresent,
  isStaged,
  isUnmerged,
  isWrappedQuote,
  objectEntries,
  objectifyPath,
  partition,
  removeDuplicates,
  removeNullableProperties,
  removeObjectProperty,
  removeUndefined,
  symmetrical,
  toArrayBuffer,
} from './utils';
export {versions} from './versions';
