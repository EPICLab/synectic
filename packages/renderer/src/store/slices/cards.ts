import type {EntityState, PayloadAction} from '@reduxjs/toolkit';
import {createEntityAdapter, createSlice} from '@reduxjs/toolkit';
import type {Card} from '@syn-types/card';
import {DateTime} from 'luxon';

export const cardAdapter = createEntityAdapter<Card>();

export const cardSlice = createSlice({
  name: 'cards',
  initialState: cardAdapter.getInitialState(),
  reducers: {
    cardAdded: cardAdapter.addOne,
    cardRemoved: cardAdapter.removeOne,
    cardUpdated: {
      reducer: (state: EntityState<Card>, action: PayloadAction<Card>) => {
        cardAdapter.upsertOne(state, action.payload);
      },
      prepare: (card: Card) => {
        return {payload: {...card, modified: DateTime.local().valueOf()}};
      },
    },
  },
});

export const {cardAdded, cardRemoved, cardUpdated} = cardSlice.actions;

export default cardSlice.reducer;
