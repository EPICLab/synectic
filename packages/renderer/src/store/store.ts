import {combineReducers, configureStore} from '@reduxjs/toolkit';
import branchesReducer from './slices/branches';
import cardsReducer from './slices/cards';
import commitsReducer from './slices/commits';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import modalsReducer from './slices/modals';
import reposReducer from './slices/repos';
import stacksReducer from './slices/stacks';
import {listenerMiddleware} from './listenerMiddleware';

export const rootReducer = combineReducers({
  branches: branchesReducer,
  cards: cardsReducer,
  commits: commitsReducer,
  filetypes: filetypesReducer,
  metafiles: metafilesReducer,
  modals: modalsReducer,
  repos: reposReducer,
  stacks: stacksReducer,
});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat([listenerMiddleware.middleware]),
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export default {store: store};
