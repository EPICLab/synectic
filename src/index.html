<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Synectic Development Environment</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>
