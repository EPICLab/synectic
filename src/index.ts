import { app, BrowserWindow, dialog, ipcMain } from 'electron';
import installExtension, { REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS } from 'electron-devtools-installer';
// This allows TypeScript to pick up the magic constant that's auto-generated by Forge's Webpack
// plugin that tells the Electron app where to look for the Webpack-bundled app code (depending on
// whether you're running in development or production).
declare const MAIN_WINDOW_WEBPACK_ENTRY: string;

// Handle creating/removing shortcuts on Windows when installing/uninstalling.
if (require('electron-squirrel-startup')) { // eslint-disable-line global-require
  app.quit();
}

const createWindow = (): void => {
  // Create the browser window.
  const mainWindow = new BrowserWindow({
    height: 1200,
    width: 1400,
    webPreferences: {
      contextIsolation: false,
      nodeIntegration: true,
      webviewTag: true
    }
  });

  // and load the index.html of the app.
  mainWindow.loadURL(MAIN_WINDOW_WEBPACK_ENTRY);

  // Open the DevTools.
  if (process.env.NODE_ENV !== 'production') {
    mainWindow.webContents.openDevTools();
  }
};

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);
app.whenReady().then(() => {
  installExtension([REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS])
    .then((name) => console.log(`Added extension: ${name}`))
    .catch((err) => console.log('An error occurred: ', err))
});

// Handle renderer requests for Electron `showOpenDialog` and returns filepath results.
ipcMain.handle('fileOpenDialog', async (_event, properties) => {
  return await dialog.showOpenDialog({ properties: [...properties, 'multiSelections'] });
});

// Handle renderer requests for Electron `showSaveDialog` and returns the results value.
ipcMain.handle('fileSaveDialog', async (_event, { defaultPath, buttonLabel, properties }) => {
  return await dialog.showSaveDialog({ defaultPath, buttonLabel, properties });
});

// Quit when all windows are closed, except on macOS. There, it's common
// for applications and their menu bar to stay active until the user quits
// explicitly with Cmd + Q.
app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow();
  }
});

// In this file you can include the rest of your app's specific main process
// code. You can also put them in separate files and import them here.
