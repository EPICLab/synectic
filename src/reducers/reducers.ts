import { combineReducers } from 'redux';
import {
  // ADD_TODO,
  // TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  VisibilityFilters,
  ActionType
} from '../actions/actions';

const { SHOW_ALL } = VisibilityFilters;

// type TodoItem = {
//   text: string;
//   completed: boolean;
// }

// type TodoState = {
//   visibilityFilter: string;
//   todos: TodoItem[];
// }

function visibilityFilter(state = SHOW_ALL, action: ActionType) {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

// function todos(state: TodoItem[] = [], action: ActionType) {
//   switch (action.type) {
//     case ADD_TODO:
//       return [
//         ...state,
//         {
//           text: action.text,
//           completed: false
//         }
//       ];
//     case TOGGLE_TODO:
//       return state.map((todo, index) => {
//         if (index === action.index) {
//           return Object.assign({}, todo, {
//             completed: !todo.completed
//           });
//         }
//         return todo;
//       });
//     default:
//       return state;
//   }
// }

const todoApp = combineReducers({
  visibilityFilter
});

export default todoApp;