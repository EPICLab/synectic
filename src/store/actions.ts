import { UUID, Repository, Card, Stack, Metafile, Filetype } from '../types';

export enum ActionKeys {
  INITIALIZE_CANVAS,
  ADD_CARD, REMOVE_CARD, UPDATE_CARD,
  ADD_STACK, REMOVE_STACK, UPDATE_STACK,
  ADD_FILETYPE, REMOVE_FILETYPE, UPDATE_FILETYPE,
  ADD_METAFILE, REMOVE_METAFILE, UPDATE_METAFILE,
  ADD_REPO, REMOVE_REPO, UPDATE_REPO
}

export type Actions =
  | InitializeCanvasAction
  | AddCardAction
  | RemoveCardAction
  | UpdateCardAction
  | AddStackAction
  | RemoveStackAction
  | UpdateStackAction
  | AddFiletypeAction
  | RemoveFiletypeAction
  | UpdateFiletypeAction
  | AddMetafileAction
  | RemoveMetafileAction
  | UpdateMetafileAction
  | AddRepoAction
  | RemoveRepoAction
  | UpdateRepoAction;

type InitializeCanvasAction = {
  type: ActionKeys.INITIALIZE_CANVAS;
};

type AddCardAction = {
  type: ActionKeys.ADD_CARD;
  id: UUID;
  card: Card;
};

type RemoveCardAction = {
  type: ActionKeys.REMOVE_CARD;
  id: UUID;
};

type UpdateCardAction = {
  type: ActionKeys.UPDATE_CARD;
  id: UUID;
  card: Partial<Card>;
};

type AddStackAction = {
  type: ActionKeys.ADD_STACK;
  id: UUID;
  stack: Stack;
}

type RemoveStackAction = {
  type: ActionKeys.REMOVE_STACK;
  id: UUID;
};

type UpdateStackAction = {
  type: ActionKeys.UPDATE_STACK;
  id: UUID;
  stack: Partial<Stack>;
};

type AddFiletypeAction = {
  type: ActionKeys.ADD_FILETYPE;
  id: UUID;
  filetype: Filetype;
};

type RemoveFiletypeAction = {
  type: ActionKeys.REMOVE_FILETYPE;
  id: UUID;
};

type UpdateFiletypeAction = {
  type: ActionKeys.UPDATE_FILETYPE;
  id: UUID;
  filetype: Partial<Filetype>;
};

type AddMetafileAction = {
  type: ActionKeys.ADD_METAFILE;
  id: UUID;
  metafile: Metafile;
};

type RemoveMetafileAction = {
  type: ActionKeys.REMOVE_METAFILE;
  id: UUID;
};

type UpdateMetafileAction = {
  type: ActionKeys.UPDATE_METAFILE;
  id: UUID;
  metafile: Partial<Metafile>;
};

type AddRepoAction = {
  type: ActionKeys.ADD_REPO;
  id: UUID;
  repo: Repository;
};

type RemoveRepoAction = {
  type: ActionKeys.REMOVE_REPO;
  id: UUID;
}

type UpdateRepoAction = {
  type: ActionKeys.UPDATE_REPO;
  id: UUID;
  repo: Partial<Repository>;
};