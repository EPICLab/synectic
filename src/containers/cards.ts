import { DateTime } from 'luxon';
import { v4 } from 'uuid';

import type { Card } from '../types';
import { ActionKeys, NarrowActionType } from '../store/actions';
import { HandlerRequiredMetafile } from './handlers';

type AddCardAction = NarrowActionType<ActionKeys.ADD_CARD>;
type UpdateCardAction = NarrowActionType<ActionKeys.UPDATE_CARD>;

/**
 * Action Creator for composing a valid ADD_CARD Redux Action.
 * @param metafile The related `Metafile` object containing a valid `handler` field.
 * @return An `AddCardAction` object that can be dispatched via Redux.
 */
export const addCard = (metafile: HandlerRequiredMetafile): AddCardAction => {
  const card: Card = {
    id: v4(),
    name: metafile.name,
    created: DateTime.local(),
    modified: metafile.modified,
    left: 10,
    top: 25,
    type: metafile.handler,
    metafile: metafile.id
  };
  return {
    type: ActionKeys.ADD_CARD,
    id: card.id,
    card: card
  };
}

/**
 * Action Creator for composing a valid UPDATE_CARD Redux Action. If the current Redux store does not contain a
 * matching card (based on UUID) for the passed parameter, then dispatching this action will not result in any
 * changes in the Redux store state.
 * @param card A `Card` object containing new field values to be updated.
 * @return An `UpdateCardAction` object that can be dispatched via Redux, including an updated timestamp in the
 * `modified` field.
 */
export const updateCard = (card: Card): UpdateCardAction => {
  return {
    type: ActionKeys.UPDATE_CARD,
    id: card.id,
    card: { ...card, modified: DateTime.local() }
  }
}