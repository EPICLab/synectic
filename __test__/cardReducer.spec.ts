import { DateTime } from 'luxon';

// import { Card, CardType } from '../src/types';
import { ActionKeys } from '../src/store/actions';
import { cardReducer } from '../src/store/reducers/cards';
import { Card } from '../src/types';

describe('cardReducer', () => {
  const cards: { [id: string]: Card } = {
    '40d14391c': {
      id: '40d14391c',
      name: 'card1',
      type: 'Editor',
      metafile: '29334943',
      created: DateTime.fromISO('2019-11-19T19:22:47.572-08:00', { setZone: true }),
      modified: DateTime.fromISO('2019-11-19T19:22:47.572-08:00', { setZone: true }),
      left: 0, top: 0
    }
  }

  const newCard: Card = {
    id: 't829w0351',
    name: 'card2',
    type: 'Editor',
    metafile: '84354571',
    created: DateTime.fromISO('2014-04-09T08:14:02.371-08:00', { setZone: true }),
    modified: DateTime.fromISO('2014-06-23T21:58:44.507-08:00', { setZone: true }),
    left: 100, top: 50
  }

  it('cardReducer appends a new card to state on action ADD_CARD', () => {
    const addedCards = cardReducer(cards, { type: ActionKeys.ADD_CARD, id: newCard.id, card: newCard });
    expect(Object.keys(cards)).toHaveLength(1);
    expect(addedCards).toMatchSnapshot();
  });

  it('cardReducer removes a card from state on action REMOVE_CARD', () => {
    const matchedCards = cardReducer(cards, { type: ActionKeys.REMOVE_CARD, id: '40d14391c' });
    expect(Object.keys(matchedCards)).toHaveLength(0);
  });

  it('cardReducer resolves non-matching card in state on action REMOVE_CARD', () => {
    const nonMatchedCards = cardReducer(cards, { type: ActionKeys.REMOVE_CARD, id: '010101010' });
    expect(Object.keys(nonMatchedCards)).toHaveLength(Object.keys(cards).length);
  });

  it('cardReducer updates state of matched card on action UPDATE_CARD', () => {
    const updatedCards = cardReducer(cards, {
      type: ActionKeys.UPDATE_CARD, id: '40d14391c', card: {
        modified: DateTime.fromISO('2019-11-22T12:54:11.374-08:00', { setZone: true }),
        left: 178,
        top: 540
      }
    });
    expect(updatedCards).not.toMatchObject(cards);
    expect(updatedCards).toMatchSnapshot();
  });
});