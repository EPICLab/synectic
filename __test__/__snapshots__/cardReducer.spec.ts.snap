// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardReducer cardReducer appends a new card to state on action ADD_CARD 1`] = `
Object {
  "40d14391c": Object {
    "created": "2019-11-19T19:22:47.572-08:00",
    "id": "40d14391c",
    "isCaptured": false,
    "left": 0,
    "metafile": "29334943",
    "modified": "2019-11-19T19:22:47.572-08:00",
    "name": "card1",
    "top": 0,
  },
  "t829w0351": Object {
    "created": "2014-04-09T09:14:02.371-07:00",
    "id": "t829w0351",
    "isCaptured": false,
    "left": 100,
    "metafile": "84354571",
    "modified": "2014-06-23T22:58:44.507-07:00",
    "name": "card2",
    "top": 50,
  },
}
`;

exports[`cardReducer cardReducer updates state of matched card on action UPDATE_CARD 1`] = `
Object {
  "40d14391c": Object {
    "created": "2019-11-19T19:22:47.572-08:00",
    "id": "40d14391c",
    "isCaptured": false,
    "left": 178,
    "metafile": "29334943",
    "modified": "2019-11-22T12:54:11.374-08:00",
    "name": "card1",
    "top": 540,
  },
}
`;
